### PR TITLE
[BUGFIX] Keep KopEventManager as a single-threaded event model

### LINKED ISSUE #756: [BUG] Keep KopEventManager as a single-threaded event model
**Describe the bug**
#735 implements tenant aware metadata topics. It instantiates a KopEventManager for each GroupCoordinator. In fact, this is wrong. KopEventManger should be a global event handler. Unless we need to use KopEventManager to implement some other models, we A single instance should be kept to handle events.

### PR DESCRIPTION
**Pull request by wenbingshen**
Fixes #756 

### Motivation
#735  implements tenant aware metadata topics. It instantiates a KopEventManager for each GroupCoordinator. In fact, this is wrong. KopEventManger should be a global event handler. Unless we need to use KopEventManager to implement some other models, we A single instance should be kept to handle events.

### Modifications
1. Keep KopEventManager as a single-threaded event model.
2. Leave a discussion on KopEventManager#307

### CODE REVIEW

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java [hunk 1]
```diff
@@ -288,54 +289,74 @@ void process(BiConsumer<String, Long> registerEventLatency,
...
+
+                    // TODO
+                    // now we have tenant/coordinator/topicsDeletions
+                    // It is possible that different coordinators have the same Kafka topics?
+                    // Because there is no multi-tenancy on the Kafka side
+                    // The question here is, which topic does the kafka admin client delete?
+                    // Is it related to tenants or delete all kafka topics with the same name under all tenants?
+                    // The implementation here only follows the concept that there is no multi-tenancy in Kafka,
+                    // that mean deleted the Kafka topics with the same name under all tenants,
+                    // and leave the rest for everyone to discuss
```
Comment (wenbingshen) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java:314]: leave a discuss

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java [hunk 2]
```diff
@@ -61,7 +62,7 @@
             new LinkedBlockingQueue<>();
     private final KopEventThread thread =
             new KopEventThread(kopEventThreadName);
-    private final GroupCoordinator coordinator;
+    private final Map<String, GroupCoordinator> groupCoordinatorsByTenant;
```
Comment (wenbingshen) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java:65]: There are no need to create a `KopEventManager` for each `GroupCoordinator`, but pass the `GroupCoordinator` list to `KopEventManager`

**Comment by wenbingshen** [comment]
ping @eolivelli @BewareMyPower @Demogorgon314 please see see it. thx

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java [hunk 3]
```diff
@@ -288,54 +289,74 @@ void process(BiConsumer<String, Long> registerEventLatency,
...
-                            kopTopic -> collectDeleteTopics.contains(kopTopic.getFullName())
-                    ).map(KopTopic::getOriginalName).collect(Collectors.toSet());
-
-                    deletedTopics.forEach(deletedTopic -> {
-                        metadataStore.delete(
-                                getDeleteTopicsPath() + "/" + deletedTopic, Optional.of((long) -1));
-                    });
-                }
+                // Localize groupCoordinatorsByTenant to avoid multi-thread conflicts
+                final Map<String, GroupCoordinator> currentCoordinators = groupCoordinatorsByTenant;
```
Comment (eolivelli) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java:300]: This is not a copy operation

**Review by eolivelli — COMMENTED** [review]
I will take a look more closer tomorrow

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java (hunk 3 above)
Comment (wenbingshen) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java:300]: You are right. I will use a copy operation to create a new Map. Thanks very much.

**Comment by wenbingshen** [comment]
Kop only exposes external topics for the Kafka client, which will be converted to the fullName pulsar topic inside kop. Maybe we need to update `KafkaRequestHandler#handleDeleteTopics`: In this method, the deleted topic will be bounded to the tenant and updated to the fullName pulsar topic topic. Another doubt is whether different tenants or different namespaces have the same Kafka topic, for example:
Kafka topic: test-topic
tenant1/namespace1/test-topic
tenant1/namespace2/test-topic
tenant2/namespace1/test-topic
tenant2/namespace3/test-topic
Is this possible?

**Comment by wenbingshen** [comment]
I found that KopTopic always uses the default tenant and namespace to create and delete topics by `public/default`,
see https://github.com/streamnative/kop/blob/master/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/utils/KopTopic.java#L76

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java [hunk 4]
```diff
@@ -288,54 +290,76 @@ void process(BiConsumer<String, Long> registerEventLatency,
...
-
-                    deletedTopics.forEach(deletedTopic -> {
-                        metadataStore.delete(
-                                getDeleteTopicsPath() + "/" + deletedTopic, Optional.of((long) -1));
-                    });
-                }
+                // Localize groupCoordinatorsByTenant to avoid multi-thread conflicts
+                final Map<String, GroupCoordinator> currentCoordinators = new HashMap<>(groupCoordinatorsByTenant);
+                final Set<String> deletedTopics = Sets.newConcurrentHashSet();
+                final AtomicInteger pendingCoordinators = new AtomicInteger(currentCoordinators.size());
```
[the remark this answers is not in the document]
Comment (wenbingshen) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java:303]: Thanks review. You are right, I want to modify it to `currentCoordinators.entrySet().stream().forEach(()->{})`. Use concurrent execution to increase the processing speed, do you think it’s okay?

[the remark this answers is not in the document]
Comment (wenbingshen) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java:303]: sorry. it should be use `currentCoordinators.entrySet().parallelStream().forEach`

[the remark this answers is not in the document]
Comment (Demogorgon314) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java:303]: In small amount of data, use concurrent may increase context switching overhead, may reduce the speed. We need to evaluate whether it is worth it. @BewareMyPower

[the remark this answers is not in the document]
Comment (wenbingshen) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java:303]: Here, If I use the int type, idea prompts to use the atomic type.

image

[the remark this answers is not in the document]
Comment (Demogorgon314) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java:303]: Oh, my mistake forget it. It still need AtomicInteger.

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java (hunk 1 above)
Comment (BewareMyPower) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java:317]: > which topic does the kafka admin client delete?

There're two configs `kafkaTenant` and `kafkaNamespace` that represent the default namespace in `KopTopic`.

Does it answer your questions? If yes, I think these comments can be removed.

Comment (wenbingshen) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java]: Yes. Let us remove these comments. :)

**Review by eolivelli — APPROVED** [review]
LGTM

**Comment by wenbingshen** [comment]
> LGTM

Thanks very much for your review. :)

**Comment by wenbingshen** [comment]
now @BewareMyPower @Demogorgon314 PTAL. Thanks.

**Review by BewareMyPower — APPROVED** [review]

**Pull request merged**